pmlc/dialect/stripe/autostencil_pass.cc

```diff
@@ -0,0 +1,312 @@
+  // Test if idx is conflict with any index in inner_idxs
+  bool ConflictInnerIndex(BlockArgument* idx);
+
+  // Number of processors
+  size_t n_cpus;
+  // The current op
+  ParallelForOp cur_op;
+  // M, N, K in inner block
+  llvm::SmallVector<BlockArgument*, numIndex> inner_idxs;
+  // M, N, K's tiles
+  size_t tiles[numIndex];
+  // Target tensors, the first two are load, the third is aggregate
+  llvm::SmallVector<Value*, numTensors> tensors;
+  // The index used by the output tensor
+  BlockArgumentSet out_idxs;
+  // The accumulation index
+  BlockArgumentSet acc_idxs;
+  // All used index
+  BlockArgumentSet all_idxs;
+  // Each index has a set of conflict index that can't be choosed into inner block together
+  llvm::DenseMap<BlockArgument*, BlockArgumentSet> conflict;
+  // The best performance
+  double best_perf;
+  // The tiles for the best performance
+  llvm::StringMap<size_t> best_tiles;
+};
+
+AutoStencil::AutoStencil(size_t processors) : n_cpus(processors) {
+}
+
+double AutoStencil::Throughput(size_t m, size_t n, size_t k) {
+  return 1;
+}
+
+double AutoStencil::Evaluate() {
+  size_t tot_inner_loop = tiles[0] * tiles[1] * tiles[1];
+  double inner_time = tot_inner_loop / Throughput(tiles[0], tiles[1], tiles[2]);
+  IVLOG(3, "Inner: loop = " << tot_inner_loop << " inner_time = " << inner_time);
+  for (size_t i = 0; i < numIndex; ++i) {
+    IVLOG(3, idxName(inner_idxs[i]).str() << ": " << tiles[i]);
+  }
+
+  llvm::DenseMap<BlockArgument*, size_t> middle_idxs;
+  for (auto idx : acc_idxs) {
+    middle_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (size_t i = 0; i < numIndex; ++i) {
+    auto it = middle_idxs.find(inner_idxs[i]);
+    if (it != middle_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  size_t tot_middle_loop = 1;
+  for (auto& kvp : middle_idxs) {
+    tot_middle_loop *= kvp.second;
+  }
+  IVLOG(3, "Middle: loop = " << tot_middle_loop);
+  for (auto& kvp : middle_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, idxName(kvp.first).str() << ": " << kvp.second);
+    }
+  }
+
+  llvm::DenseMap<BlockArgument*, size_t> outer_idxs; 
+  for (auto idx : out_idxs) {
+    outer_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (size_t i = 0; i < numIndex; ++i) {
+    auto it = outer_idxs.find(inner_idxs[i]);
+    if (it != outer_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  size_t tot_outer_loop = 1;
+  for (auto& kvp : outer_idxs) {
+    tot_outer_loop *= kvp.second;
+  }
+  IVLOG(3, "Outer: loop = " << tot_outer_loop);
+  for (auto& kvp : outer_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, idxName(kvp.first).str() << ": " << kvp.second);
+    }
+  }
+
+  size_t outer_batches = (tot_outer_loop - 1) / n_cpus + 1;
+  double perf = outer_batches * tot_middle_loop * inner_time;
+  IVLOG(3, "Performance = " << perf);
+
+  return perf;
+}
+
+void AutoStencil::SearchTiles(size_t idx) {
+  if (idx >= numIndex) {
```

Comment:
I like this search mechanism

Reply:
I'm a little worried about mutating state on the `AutoStencil` object to accumulate the tile as we descend. Have you considered passing in an array that gets populated as you go down? That way it's still 'functional' and doesn't have side effects on the `AutoStencil` object.

Reply:
We only track the 'best' output, so it's really more of a return value, but we also need the 'best cost' etc, so it would end up being some sort of composite object.  The idea that the AutoStencil object represents the search state seems Ok to me, but the fact the we reuse the AutoStencil state multiple times and clear it each time is a little awkward, so perhaps we should construct a new AutoStencil each time around the loop to make it more clear (which is basically free since it's got a mostly empty constructor).